Add support for environment variables.

jmalloc opened the pull request:
<!--
A complete guide to completing the pull request template is available at
https://github.com/dogmatiq/.github/CONTRIBUTING.md.

Don't forget to update the CHANGELOG.md file! :)
-->

#### What change does this introduce?

This PR adds support for reading environment variables as dependencies and parsing their value into various types.
It works similarly to named dependencies (see [the example](https://github.com/dogmatiq/imbue/blob/env/environmentexample_test.go) in this PR).

#### Why make this change?

It is very common to read environment variables during DI.

#### Is there anything you are unsure about?

The current implementation uses the name of the user-defined type as the name of the environment variable. This leads to the use of `SCREAMING_SNAKE_CASE` which is common for environment variables, but unusual for Go type names.

This may cause problems with some linters, and/or force the user to export the type when they may not wish to.

Another approach might be to allow `[cC]amelCaseNaming` for the type and convert that to `CAMEL_CASE_NAMING` for the call to `os.Getenv()`.

#### What issues does this relate to?

None

jmalloc commented:
This is a draft, it doesn't yet support all of the types (just `string`, `uint` and `uint16` so far).

koden-km reviewed (commented):
It looks fine to me. Should I approve or just comment since it is a draft?

danilvpetrov commented on declaration.go:
```diff
@@ -221,6 +227,10 @@ func (d *declarationOf[T]) construct(ctx *Context) error {
 		return nil
 	}
 
+	if sc, ok := any(&d.value).(selfConstructor); ok {
```
Sneaky move to convert it to `any` first and then try to type-assert it to `selfConstructor`, nice 😃

jmalloc replied:
Yeah, a bit annoying and weird and dumb, but it works haha.

danilvpetrov reviewed (approved):
I went through the entire PR and it looks really good. My only comment if we could add `float32/64` type support in case we ever need pass one.

jmalloc commented:
> It looks fine to me. Should I approve or just comment since it is a draft?

Just wanted to get your thoughts. I think this particular change is borderline "magic", my main concern is the naming thing I mentioned in the PR body.

koden-km commented on environment.go:
```diff
@@ -0,0 +1,150 @@
+// Parseable is a constraint that identifies the set of types that can be parsed
+// from their string representation, such as in an environment variable.
+type Parseable interface {
+	string |
+		int | int16 | int32 | int64 |
+		uint | uint16 | uint32 | uint64
```
Is this odd indentation something go format does?

jmalloc replied:
Yeah, it's because I've got that constraint spread across multiple lines... kind of annoying, but as they say:

> [Gofmt's style is no one's favorite, yet gofmt is everyone's favorite.](https://www.youtube.com/watch?v=PAAkCSZUG1c&t=8m43s)

jmalloc commented:
I've changed the implementation to convert the user-defined Go type name to `SCREAMING_SNAKE_CASE` in order to determine which environment variable is loaded. This way the Go identifiers still follow Go's naming rules. I've also added the remaining (useful) built-in types.

koden-km commented on internal/identifier/convert_test.go:
```diff
@@ -0,0 +1,31 @@
+	. "github.com/onsi/gomega"
+)
+
+var _ = Describe("func ToScreamingSnakeCase()", func() {
```
Does this needed a test for a single word? I did try it in the Go playground and couldn't break it though.

jmalloc replied:
Yep I'll add that test :+1:

jmalloc replied:
Done

jmalloc commented:
I'm going to merge this. I'm still a little wary of the fact that there's no longer any identifier in the code that actually matches the name of the environment variable, but we'll see how it goes.
